### PR TITLE
[SPARK-22539][SQL] Add second order for rangepartitioner since partition nu…

### PR DESCRIPTION
**Pull request by caneGuy**
…mber may be small if the specified key is skewed

## What changes were proposed in this pull request?

The rangepartitioner generated from shuffle exchange may cause partiton skew if sort key is skewed.
This patch add second order for rangepartitioner to avoid this situation.
This is an improvement from real case.

Before:
before_second

After:
after_second

## How was this patch tested?

Manually test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala [hunk 1]
```diff
@@ -193,14 +193,30 @@ object GenerateOrdering extends CodeGenerator[Seq[SortOrder], Ordering[InternalR
...
 
   def this(ordering: Seq[SortOrder], inputSchema: Seq[Attribute]) =
     this(ordering.map(BindReferences.bindReference(_, inputSchema)))
 
   @transient
-  private[this] var generatedOrdering = GenerateOrdering.generate(ordering)
+  private[this] var generatedOrdering = {
+    var generatedOrdering = GenerateOrdering.generate(ordering)
+    if (!secondOrdering.isEmpty) {
+      secondOrdering.foreach { order =>
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala]: These two lines are redundant

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala [hunk 2]
```diff
@@ -193,14 +193,30 @@ object GenerateOrdering extends CodeGenerator[Seq[SortOrder], Ordering[InternalR
...
+        try {
+          GenerateOrdering.generate(Seq(order))
+          ordering ++ Seq(order)
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala]: This does not modify a collection and it's return value goes nowhere

**Review by srowen — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala (hunk 1 above)
Comment (caneGuy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala]: Logic here used to filter input order which can not generate ordering normally.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala (hunk 2 above)
Comment (caneGuy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala]: Update @srowen  Thanks.

**Comment by hvanhovell** [comment]
Doesn't this break the `Partitioner` contract? Rows for the same key will now map to multiple partitions.

**Comment by caneGuy** [comment]
Firstly , thanks too much @hvanhovell .
And sorry for replying so late since i have some other things to handle during these time.
For the question, i think the ordering will not be broken.I did not modify the 'RangePartitioner' itself.But optimize the partitioner strategy passed from sql.And this will not break the 'Partitioner' contract since the same key will still map to the same partition.
Notice that key here contains attribute added by second order.

**Comment by caneGuy** [comment]
Ping any admin help review this?Thanks

**Comment by gczsjdy** [comment]
@caneGuy Can you give a specific example to illustrate your change? Maybe former partition result & later partition result?

**Comment by caneGuy** [comment]
@gczsjdy Ok,i will post later.

**Comment by caneGuy** [comment]
@gczsjdy Added a simple example.

**Comment by caneGuy** [comment]
Ping

**Comment by caneGuy** [comment]
Ping @srowen can you help review this?Thanks too much

**Comment by srowen** [comment]
I'm probably not qualified to review this. I don't think you addressed Herman's question. It wasn't about ordering or whether the same exact row maps to the same partition, but whether all values for a key map to the same partition. I believe that's part of the contract here. If it doesn't do that, then, I don't see how it solves the problem you're trying to solve. Skew is inherently a problem if you promise to put all values for a key together.

**Comment by hvanhovell** [comment]
@caneGuy adding the second ordering makes the sort much more fine grained, this means that the range partitioner is probably going to use different range boundaries using all the ordering parameters.

Let's work through an example here. Say we want to range partition `col1` here to 5 partitions, `col2` will be added by to the secondary ordering by your PR. `col1` has only 3 distinct values to keep the example simple. The following table shows how the columns can map to partition in the current situation and in the new situation:

| col1 | col2 | current partitioning | PR partitioning |
| ---- | ---- | ------------------- | --------------- |
| 0  | 0 | 0  | 0 |
| 1  | 0 | 1  | 1 |
| 1  | 0 | 1  | 1 |
| 1  | 1 | 1  | 2 |
| 2  | 0 | 2 | 3 |
| 2  | 1 | 2 | 4|

Notice how by adding the secondary ordering the partitioning changes. This breaks the contract of the range partitioner, and will lead to invalid results if we use this for anything besides a top level global ordering (i.e. `select * from tbl_x order by col1`).

If you only want to improve the performance of top level global orderings then it is better to write an optimizer rule/planning strategy that explicitly adds the secondary ordering columns to the ordering. The only caveat here might be that this can hurt the performance in the non-skewed; spark uses radix sorting for simple sorts which is generally faster than timsort which is used for complex cases.

**Comment by caneGuy** [comment]
Thanks @hvanhovell .IIUC, you mean that adding second order will break the contract of the range partitioner generated by `ShuffleExchange` ?If that, i agree with the opinion。
And sorry for that i can't figure out  why `select * from tbl_x order by col1` get invalid result?
May be i missed some other knowledge.Thanks again!

**Pull request closed without merging**